File: backend/app/db/models/email_marketing.py
from sqlalchemy import Integer

from .common import (
    UUID,
    Base,
    Boolean,
    Column,
    DateTime,
    ForeignKey,
    String,
    Text,
    relationship,
    utcnow,
    uuid,
)


class EmailTemplate(Base):
    """Plantilla de email reutilizable para campañas."""

    __tablename__ = "email_templates"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id = Column(UUID(as_uuid=True), ForeignKey("tenants.id"), nullable=False, index=True)
    name = Column(String(255), nullable=False)
    subject = Column(String(500), nullable=False)
    html_body = Column(Text, nullable=False)
    created_at = Column(DateTime(timezone=True), default=utcnow, nullable=False)
    updated_at = Column(DateTime(timezone=True), default=utcnow, onupdate=utcnow)

    tenant = relationship("Tenant")
    campaigns = relationship("EmailCampaign", back_populates="template")


class EmailCampaign(Base):
    """Campaña de email: conjunto de destinatarios + contenido + estado de envío."""

    __tablename__ = "email_campaigns"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id = Column(UUID(as_uuid=True), ForeignKey("tenants.id"), nullable=False, index=True)
    template_id = Column(UUID(as_uuid=True), ForeignKey("email_templates.id", ondelete="SET NULL"), nullable=True)
    name = Column(String(255), nullable=False)
    subject = Column(String(500), nullable=False)
    html_body = Column(Text, nullable=False)
    # draft | scheduled | sending | sent | failed
    status = Column(String(50), nullable=False, default="draft", index=True)
    scheduled_at = Column(DateTime(timezone=True), nullable=True, index=True)
    sent_at = Column(DateTime(timezone=True), nullable=True)
    total_count = Column(Integer, nullable=False, default=0)
    sent_count = Column(Integer, nullable=False, default=0)
    failed_count = Column(Integer, nullable=False, default=0)
    created_at = Column(DateTime(timezone=True), default=utcnow, nullable=False)
    updated_at = Column(DateTime(timezone=True), default=utcnow, onupdate=utcnow)

    tenant = relationship("Tenant")
    template = relationship("EmailTemplate", back_populates="campaigns")
    recipients = relationship("EmailCampaignRecipient", back_populates="campaign", cascade="all, delete-orphan")


class EmailCampaignRecipient(Base):
    """Destinatario individual de una campaña con su estado de envío."""

    __tablename__ = "email_campaign_recipients"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    campaign_id = Column(UUID(as_uuid=True), ForeignKey("email_campaigns.id", ondelete="CASCADE"), nullable=False, index=True)
    email = Column(String(255), nullable=False)
    name = Column(String(255), nullable=True)
    # pending | sent | failed
    status = Column(String(50), nullable=False, default="pending")
    sent_at = Column(DateTime(timezone=True), nullable=True)
    error_message = Column(Text, nullable=True)

    campaign = relationship("EmailCampaign", back_populates="recipients")

File: frontend/src/lib/api/email_marketing.ts
import { request } from "./client";

export interface EmailTemplate {
    id: string;
    name: string;
    subject: string;
    html_body: string;
    created_at: string;
    updated_at: string;
}

export interface EmailCampaign {
    id: string;
    name: string;
    subject: string;
    html_body: string;
    status: "draft" | "scheduled" | "sending" | "sent" | "failed";
    scheduled_at: string | null;
    sent_at: string | null;
    total_count: number;
    sent_count: number;
    failed_count: number;
    template_id: string | null;
    created_at: string;
}

export interface CampaignCreate {
    name: string;
    subject: string;
    html_body: string;
    template_id?: string;
    recipient_filter?: "all" | "with_email";
    scheduled_at?: string;
}

export interface TemplateCreate {
    name: string;
    subject: string;
    html_body: string;
}

export const emailMarketingApi = {
    templates: {
        list: () => request<EmailTemplate[]>("/api/v1/email-marketing/templates"),
        create: (data: TemplateCreate) =>
            request<EmailTemplate>("/api/v1/email-marketing/templates", {
                method: "POST",
                body: JSON.stringify(data),
            }),
        update: (id: string, data: Partial<TemplateCreate>) =>
            request<EmailTemplate>(`/api/v1/email-marketing/templates/${id}`, {
                method: "PUT",
                body: JSON.stringify(data),
            }),
        delete: (id: string) =>
            request<void>(`/api/v1/email-marketing/templates/${id}`, { method: "DELETE" }),
    },
    campaigns: {
        list: () => request<EmailCampaign[]>("/api/v1/email-marketing/campaigns"),
        create: (data: CampaignCreate) =>
            request<EmailCampaign>("/api/v1/email-marketing/campaigns", {
                method: "POST",
                body: JSON.stringify(data),
            }),
        update: (id: string, data: Partial<CampaignCreate>) =>
            request<EmailCampaign>(`/api/v1/email-marketing/campaigns/${id}`, {
                method: "PUT",
                body: JSON.stringify(data),
            }),
        delete: (id: string) =>
            request<void>(`/api/v1/email-marketing/campaigns/${id}`, { method: "DELETE" }),
        send: (id: string) =>
            request<{ queued: number }>(`/api/v1/email-marketing/campaigns/${id}/send`, { method: "POST" }),
        previewCount: () =>
            request<{ count: number }>("/api/v1/email-marketing/recipients/count"),
    },
};
